tls/s2n_send.c

```diff
@@ -32,6 +32,41 @@
 #include "utils/s2n_safety.h"
 #include "utils/s2n_blob.h"
 
+bool s2n_should_flush(struct s2n_connection *conn, ssize_t total_message_size, uint16_t max_write_size)
+{
+    /* If the connection is unbuffered then flush on every record written.
+     * The rest of this function assumes conn->send_mode == S2N_MULTI_RECORD_SEND */
+    if (conn->send_mode == S2N_DEFAULT_SEND) {
+        return true;
+    }
+
+    if (conn->custom_send_buffer_size < S2N_TLS_MAXIMUM_RECORD_LENGTH) {
+        return true;
+    }
+
+    uint32_t available_space = s2n_stuffer_space_remaining(&conn->out);
+    uint32_t bytes_in_stuffer = s2n_stuffer_data_available(&conn->out);
+
+    /* If the stuffer can't store the max possible max fragment size 
+     * without growing it's time to flush.
+     *
+     * We must check if conn->custom_send_buffer_size is smaller than the conn->out
+     * stuffer because is it possible that the conn->out stuffer grew larger
+     * then conn->custom_send_buffer_size. 
+     *
+     * We must also check that the stuffer has enough space for another max_write_size record. */
+    if ((bytes_in_stuffer + max_write_size > conn->custom_send_buffer_size) || available_space < max_write_size) {
+        return true;
+    }
```

Comment:
These are still the same two checks, just combined into one line.

When could the first be false but the second be true? Doesn't the second imply the first?

If you still have doubts that the first check is unnecessary, start with the second check and work backwards:
* `available_space < max_write_size` can be trivially rewritten as `max_write_size > available_space`.
* `available_space + bytes_in_stuffer = stuffer.blob.size`, so `max_write_size > (stuffer.blob.size - bytes_in_stuffer)`.
* Adding `bytes_in_stuffer` to both sides, `max_write_size + bytes_in_stuffer > stuffer.blob.size`.
* The stuffer can grow but never shrink, so `stuffer.blob.size >= conn->custom_send_buffer_size`. So `max_write_size + bytes_in_stuffer > conn->custom_send_buffer_size`.
* This can be trivially rewritten as `bytes_in_stuffer + max_write_size > conn->custom_send_buffer_size`, the first check.

Reply:
I think `available_space < max_write_size` is useful if for catching a scenario where the underlying stuffer was resized to be smaller than `conn->custom_send_buffer_size` (Which would be bad) but at least we could try to flush. What do you think?

Reply:
I'm arguing that `available_space < max_write_size` is the only useful check. The other one is unnecessary and implied by `available_space < max_write_size`. We don't care about `conn->custom_send_buffer_size`, we care about the actual size of the stuffer.

Reply:
Oh I see, that makes sense. I misunderstood the initial comment